pkg/probe/exec/exec.go

```diff
@@ -17,39 +17,102 @@ limitations under the License.
 package exec
 
 import (
+	"bytes"
+	"io"
+
+	"k8s.io/klog"
 	"k8s.io/kubernetes/pkg/probe"
 	"k8s.io/utils/exec"
+)
 
-	"k8s.io/klog"
+const (
+	maxReadBytes = 2 * 1024 // 2kb
 )
 
 // New creates a Prober.
 func New() Prober {
-	return execProber{}
+	return execProber{
+		maxReadBytes: maxReadBytes,
+	}
 }
 
 // Prober is an interface defining the Probe object for container readiness/liveness checks.
 type Prober interface {
 	Probe(e exec.Cmd) (probe.Result, string, error)
 }
 
-type execProber struct{}
+type execProber struct {
+	maxReadBytes int64
+}
 
 // Probe executes a command to check the liveness/readiness of container
 // from executing a command. Returns the Result status, command output, and
 // errors if any.
 func (pr execProber) Probe(e exec.Cmd) (probe.Result, string, error) {
-	data, err := e.CombinedOutput()
-	klog.V(4).Infof("Exec probe response: %q", string(data))
+	mr, err := combinedCmdOutput(e)
```

Comment:
If you look at the implementation of [CombinedOutput](https://golang.org/src/os/exec/exec.go?s=14425:14471#L505), you can see that it just sets `e.Stdout` & `e.Stderr` to a bytes.Buffer. What if instead, you just set those to a limited buffer, and call Run? We used to have a utility for this, [LimitWriter](https://github.com/kubernetes/kubernetes/blob/release-1.12/pkg/util/limitwriter/limitwriter.go), but it was cleaned up in https://github.com/kubernetes/kubernetes/pull/69790. It might be worth resurrecting for this.

Reply:
I think the main problem is, that in ...

https://github.com/kubernetes/kubernetes/blob/30c7df5cd822067016640aa267714204ac089172/pkg/kubelet/kuberuntime/kuberuntime_container.go#L783-L790

... `ExecSync` is called, which already returns `[]byte`s for `Std{out,err}`. That means that those `[]byte`s are already allocated in memory and even when we only use partial data from those, the underlaying problem does not go away: We have already filled the memory on the runtime side, even before this function on the kubelet side returns.
I think this needs to be changed to do an async `Exec`. I would assume that we get access to `Std{out,err}` via `io.Reader`s or such which we can close off once we have read (more than) enough data, maybe even actively kill the exec in the container.

I think this all could implemented in the above `RunInContainer` method, we'd just need to pass in limits for `Stdout` & `Stderr` (or a combined one) and handle that limiting and  back-pressure in there.

Or, alternatively, we could also change the interface of `RunInContainer` to return `io.Reader`s and to be used by the caller(s). They can then decide on how to read / close those.

May main problem is to understand how to properly use `Exec`, `ExecResponse`, in general the CRI protobuf API to properly stream in `Std{out,err}`. If you could point me to an example of the kubelet running some async command in the container and streaming out data, that'd be great! I'll also have another look.
Also, of course, if I am completely off and took a wrong turn somewhere in all the abstractions, please let me know too! :)

Reply:
Good point, this is really something that needs to be addressed on the runtime side. I suggest we add a "max_output" (or something like that) field to the `ExecSyncRequest`, and push the limiting down into the runtime.

Actually, on further investigation it looks like gRPC already enforces a max message_length, which defaults to 4MB. We should verify that we're using that, but still update the runtimes to cap that.

/cc @haiyanmeng @Random-Liu @yujuhong

Reply:
Having the limit set on the kubelet side (whether via grpc or some other mechanism) makes sense to me, since it protects kubelet even when runtime is misconfigured.

Reply:
The grpc client connecting to CRI in kubelet sets the maximum message size the client can receive to be `16MB` (maxMsgSize):
https://github.com/kubernetes/kubernetes/blob/b67126e350fdba6d89a0ee64310cac7db14e33d3/pkg/kubelet/remote/remote_runtime.go#L62